network/surveyer.go

```diff
@@ -0,0 +1,178 @@
+	"github.com/filecoin-project/lily/tasks/survey/peeragents"
+	logging "github.com/ipfs/go-log/v2"
+)
+
+const (
+	PeerAgentsTask = "peeragents" // task that observes connected peer agents
+)
+
+var log = logging.Logger("visor/network")
+
+type API interface {
+	peeragents.API
+}
+
+type SurveyerOpt func(t *Surveyer)
+
+func NewSurveyer(api API, storage model.Storage, interval time.Duration, name string, tasks []string, options ...SurveyerOpt) (*Surveyer, error) {
+	if interval <= 0 {
+		return nil, xerrors.Errorf("surveyer interval must be greater than zero: %d", interval)
+	}
+
+	obs := &Surveyer{
+		interval: interval,
+		storage:  storage,
+		name:     name,
+		tasks:    map[string]Task{},
+	}
+
+	for _, task := range tasks {
+		switch task {
+		case PeerAgentsTask:
+			obs.tasks[PeerAgentsTask] = peeragents.NewTask(api)
+		default:
+			return nil, xerrors.Errorf("unknown task: %s", task)
+		}
+	}
+
+	for _, opt := range options {
+		opt(obs)
+	}
+
+	return obs, nil
+}
+
+// An Surveyer observes features of the filecoin network
+type Surveyer struct {
+	interval time.Duration
+	storage  model.Storage
+	name     string
+	tasks    map[string]Task
+}
+
+// Run starts observing the filecoin netwoirk and continues until the context is done or
+// a fatal error occurs.
+func (s *Surveyer) Run(ctx context.Context) error {
+	ticker := time.NewTicker(s.interval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-ticker.C:
+			s.Tick(ctx)
+		}
+	}
+}
+
+func (s *Surveyer) Details() (string, map[string]interface{}) {
+	return "surveyer", map[string]interface{}{
+		"name":     s.name,
+		"interval": s.interval,
+	}
+}
+
+// Tick is called on each tick of the surveyer's interval
+func (s *Surveyer) Tick(ctx context.Context) error {
+	ctx, span := global.Tracer("").Start(ctx, "Surveyer.Tick")
+	defer span.End()
+
+	// Do as much indexing as possible in the specified time interval
+	tctx, cancel := context.WithTimeout(ctx, s.interval)
+	defer cancel()
+
+	inFlight := 0
+	results := make(chan *TaskResult, len(s.tasks))
+
+	// Run each tipset processing task concurrently
+	for name, task := range s.tasks {
+		inFlight++
+		go s.runTask(tctx, task, name, results)
+	}
+
+	// Wait for all tasks to complete
+	for inFlight > 0 {
+		var res *TaskResult
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case res = <-results:
+		}
+		inFlight--
+
+		llt := log.With("task", res.Task)
+
+		// Was there a fatal error?
+		if res.Error != nil {
+			llt.Errorw("task returned with error", "error", res.Error.Error())
+			return res.Error
+		}
+
+		llt.Debugw("task report", "time", res.CompletedAt.Sub(res.StartedAt))
+
+		startPersist := time.Now()
+		if err := s.storage.PersistBatch(ctx, res.Data); err != nil {
+			stats.Record(ctx, metrics.PersistFailure.M(1))
+			llt.Errorw("persistence failed", "error", err)
+		} else {
+			llt.Debugw("task data persisted", "time", time.Since(startPersist))
+		}
```

Comment:
I think logging is fine at moment. I don't expect any processing report mechanism since we're not processing tipsets. We're just recording the state of the network at some point in time.